### PR TITLE
[FIX] l10n_sa_edi: add support for global discounts

### PR DESCRIPTION
**Pull request by mebe-odoo**
This commit does three things:
    
1) Global Discounts: When creating an invoice with a negative line, such as a Tax Invoice with global discount, we process the line as a document level allowance as required by ZATCA, and remove it from the lines sent in the UBL. This allows us to provide support for global discounts, such as those applied by promotion programs

2) Add support for down-payment document type in _l10n_sa_get_invoice_type method: 386
    
3) Trim spaces from phone numbers before submission to avoid receiving warnings from ZATCA

Description of the issue/feature this PR addresses:
Process global discounts into AllowanceCharge values for ZATCA saudi eInvoicing.
Avoid warnings regarding spaces in phone numbers.
Add support for down-payment invoice type code.

Current behavior before PR:
If a promotions program is applied on a sale order or POS order, or a line with a negative amount is added on an invoice, the line is sent as is to the authorities, which triggers an error as ZATCA only accepts positive amount lines.
If a phone number is provided, ZATCA returns a warning regarding the phone number format.

Desired behavior after PR is merged:
Negative lines on an invoice, that are not related to a down-payment, should be processed as document level allowances.
Phone numbers are correctly formatted prior to sending to ZATCA.
Down-payment invoices use the appropriate document type code.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_sa_edi/__manifest__.py [hunk 1]
```diff
@@ -31,7 +31,7 @@
         'views/res_partner_views.xml',
         'views/res_company_views.xml',
         'views/res_config_settings_view.xml',
-        'views/report_invoice.xml',
+        'views/report_invoice.xml'
```
Comment (jco-odoo) [inline addons/l10n_sa_edi/__manifest__.py:34]: unnecessary diff

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py [hunk 2]
```diff
@@ -115,6 +115,11 @@ def _get_invoice_payment_means_vals_list(self, invoice):
         res[0]['adjustment_reason'] = invoice.ref
         return res
 
+    def _get_partner_contact_vals(self, partner):
```
Comment (jco-odoo) [inline addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:118]: please mention in the commit that we added some backport

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py [hunk 3]
```diff
@@ -155,7 +160,11 @@ def _l10n_sa_get_invoice_type(self, invoice):
                 - 381: Credit Note
                 - 388: Invoice
         """
-        return 383 if invoice.debit_origin_id else 381 if invoice.move_type == 'out_refund' else 388
+        return (
+            383 if invoice.debit_origin_id else
+            381 if invoice.move_type == 'out_refund' else
```
Comment (jco-odoo) [inline addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:165]: same?

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py [hunk 4]
```diff
@@ -209,37 +218,60 @@ def _l10n_sa_get_monetary_vals(self, invoice, vals):
...
-            # Makes no sense, but according to ZATCA, if there is a downpayment, the TotalInclusiveAmount
-            # should include the total amount of the invoice (including downpayment amount) PLUS the downpayment
-            # total amount, AGAIN.
-            prepaid_amount = tax_inclusive_amount + downpayment_vals['total_amount']
-            payable_amount = - downpayment_vals['total_amount']
-
+            prepaid_amount = downpayment_vals['total_amount']
+            payable_amount -= prepaid_amount
         return {
+            'line_extension_amount': line_extension_amount - allowance_total_amount,
```
Comment (jco-odoo) [inline addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:237]: I wonder if that should not be in generic then?

**Review by jco-odoo — COMMENTED** [review]
Why do we need changes to the line discount?

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py (hunk 4 above)
Comment (mebe-odoo) [reply addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:237]: It should be, indeed

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py [hunk 5]
```diff
@@ -268,9 +300,21 @@ def _export_invoice_vals(self, invoice):
...
         return vals
 
+    def _l10n_sa_postprocess_line_vals(self, vals):
+        """
+            Postprocess vals to remove negative line amounts, as those will be used to compute
+            document level allowances (global discounts)
+        """
+        final_line_vals = []
+        for line_vals in vals['vals']['invoice_line_vals']:
+            if line_vals['price_vals']['price_amount'] >= 0:
```
[the remark this answers is not in the document]
Comment (mebe-odoo) [reply addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:314]: Yep, a line with a negative amount (price or qty) will be rejected

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py (hunk 2 above)
Comment (mebe-odoo) [reply addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:118]: done

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py [hunk 6]
```diff
@@ -209,37 +218,60 @@ def _l10n_sa_get_monetary_vals(self, invoice, vals):
...
         # - The payable amount does not account for the tax amount of the downpayment, so we add it
         downpayment_vals = self._l10n_sa_get_prepaid_amount(invoice, vals)
-
+        allowance_charge_vals = vals['vals']['allowance_charge_vals']
+        allowance_total_amount = sum(a['amount'] for a in allowance_charge_vals if a['charge_indicator'] == 'false')
         if downpayment_vals:
-            # Makes no sense, but according to ZATCA, if there is a downpayment, the TotalInclusiveAmount
-            # should include the total amount of the invoice (including downpayment amount) PLUS the downpayment
-            # total amount, AGAIN.
-            prepaid_amount = tax_inclusive_amount + downpayment_vals['total_amount']
```
Comment (jco-odoo) [inline addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:228]: tax_inclusive_amount not needed anymore?

**Review by jco-odoo — COMMENTED** [review]
Please check this still.

File: addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py (hunk 6 above)
Comment (mebe-odoo) [reply addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py:228]: I updated the computation logic to be more in lign with rules BR-KSA-80 and and BR-CO-16. I double tested this for standard invoices, downpayments & invoices, and it checks out

**Comment by jco-odoo** [comment]
@robodoo r+
